Adding Redaction and Compliance docs

mariamgerges opened the pull request:
This PR adds documentation pages for redaction and compliance.
<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/extensions/compliance.md](https://github.com/dotnet/docs/blob/e077d252827e49ea1addb19b1df7141f0a76927f/docs/core/extensions/compliance.md) | [docs/core/extensions/compliance](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/compliance?branch=pr-en-us-45373) |
| [docs/core/extensions/data-classification.md](https://github.com/dotnet/docs/blob/e077d252827e49ea1addb19b1df7141f0a76927f/docs/core/extensions/data-classification.md) | [docs/core/extensions/data-classification](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/data-classification?branch=pr-en-us-45373) |
| [docs/core/extensions/data-redaction.md](https://github.com/dotnet/docs/blob/e077d252827e49ea1addb19b1df7141f0a76927f/docs/core/extensions/data-redaction.md) | [docs/core/extensions/data-redaction](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/data-redaction?branch=pr-en-us-45373) |
| [docs/fundamentals/toc.yml](https://github.com/dotnet/docs/blob/e077d252827e49ea1addb19b1df7141f0a76927f/docs/fundamentals/toc.yml) | [docs/fundamentals/toc](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/toc?branch=pr-en-us-45373) |


<!-- PREVIEW-TABLE-END -->

Commit:
Docs update

Commit:
Docs update

dariusclay commented:
@IEvangelist this is for the data classification and redaction extensions in dotnet/extensions. Thank you @mariamgerges for taking this on! 

Question I have, is location within the toc -- I would expect it to be under [Runtime Libraries Overview](https://learn.microsoft.com/en-us/dotnet/standard/runtime-libraries-overview), but want to hear some opinions from the folks here.

IEvangelist commented:
> @IEvangelist this is for the data classification and redaction extensions in dotnet/extensions. Thank you @mariamgerges for taking this on!
> 
> Question I have, is location within the toc -- I would expect it to be under [Runtime Libraries Overview](https://learn.microsoft.com/en-us/dotnet/standard/runtime-libraries-overview), but want to hear some opinions from the folks here.

Hi @dariusclay - yeah, I agree it should be under the Runtime libraries TOC instead. This is neither diagnostics nor instrumentation. I'll review this in a bit. Great job @mariamgerges

Commit:
moving docs

Commit:
Merge branch 'main' into mariamaziz/redaction-compliance-docs

Commit:
moving docs

Commit:
doc update

Commit:
doc update

Commit:
doc update

mariamgerges commented:
@IEvangelist The OPS status checker build step is failing, and I don't understand why, can you help?

IEvangelist commented:
> @IEvangelist The OPS status checker build step is failing, and I don't understand why, can you help?

It's a safety net, we have that check in place to fail when the [OpenPublishing.Build](https://buildapi.docs.microsoft.com/Output/PullRequest/a3bda507-3390-de91-8f7f-26f90f4e5fc8/202503191227031487-45373/BuildReport?accessString=f32b35361601ee97570b7d1b8438be10ad8e2f9e10cc0d9f19452e276f4a2bae) check has warnings. I don't think any of these are related to your PR, so you're good.

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+
+Redaction helps you sanitize or mask sensitive info in logs, error messages, or other outputs. This keeps you compliant with privacy rules and protects sensitive data. It's useful in apps that handle personal data, financial info, or other confidential info.
+
+## Install the package
```
```suggestion
## Install redaction package
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,135 @@
+
+## Install the package
+
+From the command line:
+
+```console
+dotnet add package Microsoft.Extensions.Compliance.Redaction
+```
+
+Or directly in the C# project file:
+
+```xml
+<ItemGroup>
+  <PackageReference Include="Microsoft.Extensions.Compliance.Redaction" Version="[CURRENTVERSION]" />
+</ItemGroup>
+```
````
Same recommendation as mentioned earlier, about tabs here.

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,135 @@
+</ItemGroup>
+```
+
+## Redactors
````
```suggestion
## Available redactors
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+
+## Redactors
+
+Redactors handle the actual redaction of sensitive data. They replace or mask sensitive information.
```
```suggestion
Redactors are responsible for the act of redacting sensitive data. They redact, replace, or mask sensitive information. Consider the following available redactors provided by the library:
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+- The <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactor> uses `HMACSHA256` to encode data being redacted.
+
+## Usage example
+
+### Register the services
+
+Register the services using one of the `AddRedaction` methods:
```
```suggestion

To use the built-in redactors, you have to register the required services. Register the services using one of the available `AddRedaction` methods as described in the following list:
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,135 @@
+
+Register the services using one of the `AddRedaction` methods:
+
+```csharp
+public static IServiceCollection AddRedaction(this IServiceCollection services)
+public static IServiceCollection AddRedaction(this IServiceCollection services, Action<IRedactionBuilder> configure)
+```
````
Remove the code section here, and replace it with the following unordered list:

```markdown
- <xref:Microsoft.Extensions.DependencyInjection.RedactionServiceCollectionExtensions.AddRedaction(Microsoft.Extensions.DependencyInjection.IServiceCollection)>: Registers an implementation of <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider> in the <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>.
- <xref:Microsoft.Extensions.DependencyInjection.RedactionServiceCollectionExtensions.AddRedaction(Microsoft.Extensions.DependencyInjection.IServiceCollection,System.Action{Microsoft.Extensions.Compliance.Redaction.IRedactionBuilder})>: Registers an implementation of <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider> in the <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> and configures available redactors with the given `configure` delegate.
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+{
+    public static string Name => "MyTaxonomy";
+
+    public static DataClassification Private => new DataClassification(Name, nameof(Private));
+    public static DataClassification Public => new DataClassification(Name, nameof(Public));
+    public static DataClassification Personal => new DataClassification(Name, nameof(Personal));
```
```suggestion
    public static DataClassification Private => new(Name, nameof(Private));
    public static DataClassification Public => new(Name, nameof(Public));
    public static DataClassification Personal => new(Name, nameof(Personal));
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+
+### Configure a redactor
+
+Fetch redactors at runtime using an `IRedactorProvider`. You can implement your own provider and register it inside the `AddRedaction` call, or use the default provider. Configure redactors using these `IRedactionBuilder` methods:
```
```suggestion
Fetch redactors at runtime using an <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider>. You can implement your own provider and register it inside the `AddRedaction` call, or use the default provider. Configure redactors using these <xref:Microsoft.Extensions.Compliance.Redaction.IRedactionBuilder> methods:
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,135 @@
+Configure the HMAC redactor using these `IRedactionBuilder` methods:
+
+```csharp
+var serviceCollection = new ServiceCollection();
````
I'm confused as to why we're manually instantiating a `ServiceCollection` in this example, but in the earlier example, we were using a `builder`. I'd recommend that we try to stay consistent with this.

mariamgerges replied:
I updated the previous example to be instantiating a ServiceCollection like this example for consistency.

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,135 @@
+
+```csharp
+// This will use the default redactor, which is the ErasingRedactor
+builder.Services.AddRedaction();
````
Where does this `builder` instance come from? If we're going to not show that code, we should at least mention when type this builder is-I assume it's an `IHostApplicationBuilder`.

mariamgerges replied:
I updated this to be using a serviceCollection instead of builder instance, so the document would be consistent.

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,135 @@
+}
+```
+
+- The `HmacRedactorOptions` requires its `KeyId` and `Key` properties to be set.
````
```suggestion
- The <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactorOptions> requires its <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactorOptions.Key?displayProperty=nameWithType> and <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactorOptions.KeyId?displayProperty=nameWithType> properties to be set.
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+- The `KeyId` is appended to each redacted value to identify the key used to hash the data.
+- Different key IDs mean the values are unrelated and can't be used for correlation.
+
+> [!Note]
```
```suggestion
> [!NOTE]
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,135 @@
+- Different key IDs mean the values are unrelated and can't be used for correlation.
+
+> [!Note]
+> The `HmacRedactor` is still experimental, so the above methods will show warning `EXTEXP0002` indicating it's not yet stable.
```
```suggestion
> The <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactor> is still experimental, so the preceding methods will cause the `EXTEXP0002` warningm indicating it's not yet stable.
```

IEvangelist reviewed (commented):
This is looking great, lots of nits to help improve this already awesome content. Let's get these suggestions in and I'll do another small review.

mariamgerges commented on docs/core/extensions/compliance.md:
````diff
@@ -0,0 +1,190 @@
+```csharp
+using Microsoft.Extensions.Compliance.Classification;
+
+public static class MyTaxonomyClassifications
````
I think they can be internal.

mariamgerges replied:
I think we can keep it internal and add a note explaining your point.

mariamgerges commented on docs/core/extensions/compliance.md:
````diff
@@ -0,0 +1,190 @@
+Example:
+
+```csharp
+using Microsoft.Extensions.Compliance.Classification;
````
sorry, I don't get your question.

IEvangelist replied:
The namespace used here is an official Microsoft namespace, I know there's conflicting opinions about that usage. Some think unless you're a Microsoft project, you shouldn't use this namespace for your custom code, while others think it's fine. Since this code is going into docs, it implies that many developers will be simply copying and pasting it, resulting in tons of custom code sharing this namespace. The choice is yours.

dariusclay replied:
Ah, I see the confusion. The namespace is the using, which comes from the package they need to reference for DataClassification.

dariusclay replied:
Same with rest of the examples, there's no namespace defined just using statements.

mariamgerges replied:
ah yes, that's why I got confused too

IEvangelist replied:
Ah, sorry. I read that wrong - my bad.

Commit:
doc update based on suggestions

Commit:
doc update

Commit:
doc update

Commit:
doc update

Commit:
doc update

Commit:
doc update

IEvangelist committed:
Improve formatting and add example class in compliance doc

IEvangelist committed:
Add missing section and code example in docs

IEvangelist commented:
This is really shaping up nicely. I think we might want to make a few more tweaks. We could benefit from an Overview article that speaks to the benefits of these libraries and their overall compliance objectives. Then, in the compliance article, let's pivot a bit, moving the Redaction bits into the redaction article, and shifting this to be more about data classification. Then the TOC would look more like this:

- Compliance
  - Overview
  - Data classification
  - Redaction

Commit:
doc update

Commit:
doc fix

Commit:
doc fix

Commit:
doc fix

mariamgerges commented:
> This is really shaping up nicely. I think we might want to make a few more tweaks. We could benefit from an Overview article that speaks to the benefits of these libraries and their overall compliance objectives. Then, in the compliance article, let's pivot a bit, moving the Redaction bits into the redaction article, and shifting this to be more about data classification. Then the TOC would look more like this:
> 
> * Compliance
>   
>   * Overview
>   * Data classification
>   * Redaction

I updated the docs, created an overview article and a separate doc for data classification and moved the redaction bits to the data redaction doc and updated the TOC

IEvangelist commented on docs/core/extensions/data-classification.md:
````diff
@@ -0,0 +1,128 @@
+}
+```
+
+If you want to share your custom classification taxonomy among your applications, this class should be public. For example, you can have a shared library containing custom classifications, that you can use in multiple applications.
````
```suggestion
If you want to share your custom classification taxonomy with other apps, this class and its members should be `public` instead of `internal`. For example, you can have a shared library containing custom classifications, that you can use in multiple applications.
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,176 @@
+---
+title: Data redaction in .NET
+description: Learn how to use .NET data redaction libraries to protect your application's sensitive data.
+ms.date: 03/17/2025
```
```suggestion
ms.date: 03/20/2025
```

mariamgerges replied:
done

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,176 @@
+Redaction helps you sanitize or mask sensitive information in logs, error messages, or other outputs. This keeps you compliant with privacy rules and protects sensitive data. It's useful in apps that handle personal data, financial information, or other confidential data points.
+
+## Install redaction package
+
```
```suggestion

To get started, install the [📦 Microsoft.Extensions.Compliance.Redaction](https://www.nuget.org/packages/Microsoft.Extensions.Compliance.Redaction) NuGet package:

```

IEvangelist commented on docs/core/extensions/data-redaction.md:
````diff
@@ -0,0 +1,176 @@
+```csharp
+using Microsoft.Extensions.Compliance.Classification;
+
+public static class MyTaxonomy
````
```suggestion
public static class MyTaxonomyClassifications
```

mariamgerges replied:
done

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,176 @@
+
+### Configure the HMAC redactor
+
+Configure the HMAC redactor using these `IRedactionBuilder` methods:
```
```suggestion
Configure the HMAC redactor using these <xref:Microsoft.Extensions.Compliance.Redaction.IRedactionBuilder> methods:
```

IEvangelist commented on docs/core/extensions/data-redaction.md:
```diff
@@ -0,0 +1,176 @@
+{
+    private static readonly StarRedactor _starRedactor = new();
+
+    public static StarRedactorProvider Instance { get; } = new();
```
Is the `Instance` member needed? Also, for the `GetRedactor` API when we implement a custom one like shown here, when should we consider the given `DataClassificationSet` - what's the used for?

IEvangelist reviewed (commented):
Looking much better, we have a few more things to improve then we can ship this!

Commit:
docs updates

Commit:
docs updates
